File: src/main/scala/com/vc/easy/L501.scala
package com.vc.easy

class TreeNode501(var _value: Int) {
  var value: Int = _value
  var left: TreeNode501 = null
  var right: TreeNode501 = null
}

object L501 {
  def findMode(root: TreeNode501): Array[Int] = {
    import scala.collection.mutable
    val st = new mutable.Stack[TreeNode501]()
    val map = mutable.Map[Int, Int]()
    var current = root
    while(current != null || st.nonEmpty) {
      if(current != null) {
        map.put(current.value, map.getOrElse(current.value, 0) + 1)
        st.push(current)
        current = current.left
      }
      else current = st.pop.right
    }
    if(map.isEmpty) return Array()
    val maxCnt = map.maxBy(_._2)._2
    map.filter(_._2 == maxCnt).keys.toArray
  }
}

File: src/main/java/com/vc/easy/L501.java
package com.vc.easy;

import com.vc.hard.TreeNode;

import java.util.*;

class L501 {
    private int maxCount = 0, currentCount = 1;
    private List<Integer> modes;
    private Integer prev;
    public int[] findMode(TreeNode root) {
        this.modes = new ArrayList<>();
        inorder(root);
        int[] res = new int[modes.size()];
        for(int i = 0; i < modes.size(); i++) res[i] = modes.get(i);
        return res;
    }

    private void inorder(TreeNode root) {
        if(root == null) return;
        inorder(root.left);
        if(prev != null) {
            if(prev == root.val) currentCount++;
            else currentCount = 1;
        }
        if(currentCount > maxCount) {
            maxCount = currentCount;
            modes.clear();
            currentCount = maxCount;
            modes.add(root.val);
        }
        else if(currentCount == maxCount) modes.add(root.val);
        prev = root.val;
        inorder(root.right);
    }
}
